File: src/sys/java/fan/sys/MimeType.java
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   14 May 08  Brian Frank  Creation
//

package fan.sys;

import java.util.HashMap;
import java.util.Map.Entry;
import java.util.Iterator;

/**
 * MimeType represents the parsed value of a Content-Type
 * header per RFC 2045 section 5.1.
 */
public final class MimeType
  extends FanObj
{

//////////////////////////////////////////////////////////////////////////
// Construction
//////////////////////////////////////////////////////////////////////////

  public static MimeType fromStr(String s) { return fromStr(s, true); }
  public static MimeType fromStr(String s, boolean checked)
  {
    try
    {
      // common interned mime types
      switch (s.charAt(0))
      {
        case 'i':
          if (s.equals("image/png"))  return imagePng;
          if (s.equals("image/jpeg")) return imageJpeg;
          if (s.equals("image/gif"))  return imageGif;
          break;
        case 't':
          if (s.equals("text/plain")) return textPlain;
          if (s.equals("text/html"))  return textHtml;
          if (s.equals("text/xml"))   return textXml;
          if (s.equals("text/plain; charset=utf-8")) return textPlainUtf8;
          if (s.equals("text/html; charset=utf-8"))  return textHtmlUtf8;
          if (s.equals("text/xml; charset=utf-8"))   return textXmlUtf8;
          break;
        case 'x':
          if (s.equals("x-directory/normal")) return dir;
          break;
      }

      // parse
      return parse(s);
    }
    catch (ParseErr e)
    {
      if (!checked) return null;
      throw e;
    }
    catch (Exception e)
    {
      if (!checked) return null;
      throw ParseErr.make("MimeType",  s);
    }
  }

  private static MimeType parse(String s)
  {
    int slash = s.indexOf('/');
    String media = s.substring(0, slash);
    String sub = s.substring(slash+1, s.length());
    Map params = emptyParams();

    int semi = sub.indexOf(';');
    if (semi > 0)
    {
      params = doParseParams(sub, semi+1);
      sub = sub.substring(0, semi).trim();
    }

    MimeType r  = new MimeType();
    r.str       = s;
    r.mediaType = FanStr.lower(media);
    r.subType   = FanStr.lower(sub);
    r.params    = params.ro();
    return r;
  }

  public static Map parseParams(String s) { return parseParams(s, true); }
  public static Map parseParams(String s, boolean checked)
  {
    try
    {
      return doParseParams(s, 0);
    }
    catch (ParseErr e)
    {
      if (!checked) return null;
      throw e;
    }
    catch (Exception e)
    {
      if (!checked) return null;
      throw ParseErr.make("MimeType params",  s);
    }
  }

  private static Map doParseParams(String s, int offset)
  {
    Map params = new Map(Sys.StrType, Sys.StrType);
    params.caseInsensitive(true);
    boolean inQuotes = false;
    int keyStart = offset;
    int valStart = -1;
    int valEnd   = -1;
    int eq       = -1;
    boolean hasEsc = false;
    for (int i = keyStart; i<s.length(); ++i)
    {
      int c = s.charAt(i);

      if (c == '(' && !inQuotes)
        throw ParseErr.make("MimeType", s, "comments not supported");

      if (c == '=' && !inQuotes)
      {
        eq = i++;
        while (FanInt.isSpace(s.charAt(i))) ++i;
        if (s.charAt(i) == '"') { inQuotes = true; ++i; }
        else inQuotes = false;
        valStart = i;
      }

      if (eq < 0) continue;

      if (c == '\\' && inQuotes)
      {
        ++i;
        hasEsc = true;
        continue;
      }

      if (c == '"' && inQuotes)
      {
        valEnd = i-1;
        inQuotes = false;
      }

      if (c == ';' && !inQuotes)
      {
        if (valEnd < 0) valEnd = i-1;
        String key = s.substring(keyStart, eq).trim();
        String val = s.substring(valStart, valEnd+1).trim();
        if (hasEsc) val = unescape(val);
        params.set(key, val);
        keyStart = i+1;
        eq = valStart = valEnd = -1;
        hasEsc = false;
      }
    }

    if (keyStart < s.length())
    {
      if (valEnd < 0) valEnd = s.length()-1;
      String key = s.substring(keyStart, eq).trim();
      String val = s.substring(valStart, valEnd+1).trim();
      if (hasEsc) val = unescape(val);
      params.set(key, val);
    }

    return params;
  }

  private static String unescape(String s)
  {
    StringBuilder buf = new StringBuilder(s.length());
    for (int i=0; i<s.length(); ++i)
    {
      int c = s.charAt(i);
      if (c != '\\') buf.append((char)c);
      else if (s.charAt(i+1) == '\\') { buf.append('\\'); i++; }
    }
    return buf.toString();
  }

//////////////////////////////////////////////////////////////////////////
// Extension
//////////////////////////////////////////////////////////////////////////

  public static MimeType forExt(String s)
  {
    if (s == null) return null;
    try
    {
      String val = (String)Sys.sysPod.props(etcUri, Duration.oneMin).get(FanStr.lower(s));
      if (val == null) return null;
      return MimeType.fromStr(val);
    }
    catch (Exception e)
    {
      System.out.println("MimeType.forExt: " + s);
      e.printStackTrace();
      return null;
    }
  }

  static final Uri etcUri = Uri.fromStr("ext2mime.props");

//////////////////////////////////////////////////////////////////////////
// Identity
//////////////////////////////////////////////////////////////////////////

  public boolean equals(Object obj)
  {
    if (!(obj instanceof MimeType)) return false;
    MimeType x = (MimeType)obj;
    return mediaType.equals(x.mediaType) &&
           subType.equals(x.subType) &&
           params.equals(x.params);
  }

  public int hashCode()
  {
    return mediaType.hashCode() ^
           subType.hashCode() ^
           params.hashCode();
  }

  public long hash()
  {
    return hashCode();
  }

  public String toStr()
  {
    return str;
  }

  public Type typeof()
  {
    return Sys.MimeTypeType;
  }

//////////////////////////////////////////////////////////////////////////
// Methods
//////////////////////////////////////////////////////////////////////////

  public String mediaType()
  {
    return mediaType;
  }

  public String subType()
  {
    return subType;
  }

  public Map params()
  {
    return params;
  }

  public Charset charset()
  {
    String s = (String)params().get("charset");
    if (s == null) return Charset.utf8;
    return Charset.fromStr(s);
  }

//////////////////////////////////////////////////////////////////////////
// Lazy Load
//////////////////////////////////////////////////////////////////////////

  static Map emptyParams()
  {
    Map q = emptyQuery;
    if (q == null)
    {
      q = new Map(Sys.StrType, Sys.StrType);
      q.caseInsensitive(true);
      q = (Map)q.toImmutable();
      emptyQuery = q;
    }
    return q;
  }
  static Map emptyQuery;

//////////////////////////////////////////////////////////////////////////
// Fields
//////////////////////////////////////////////////////////////////////////

  static final MimeType imagePng   = parse("image/png");
  static final MimeType imageGif   = parse("image/gif");
  static final MimeType imageJpeg  = parse("image/jpeg");
  static final MimeType textPlain  = parse("text/plain");
  static final MimeType textHtml   = parse("text/html");
  static final MimeType textXml    = parse("text/xml");
  static final MimeType dir        = parse("x-directory/normal");
  static final MimeType textPlainUtf8 = parse("text/plain; charset=utf-8");
  static final MimeType textHtmlUtf8  = parse("text/html; charset=utf-8");
  static final MimeType textXmlUtf8   = parse("text/xml; charset=utf-8");

  private String mediaType;
  private String subType;
  private Map params;
  private String str;

}

File: src/sys/dotnet/fan/sys/MimeType.cs
//
// Copyright (c) 2008, Brian Frank and Andy Frank
// Licensed under the Academic Free License version 3.0
//
// History:
//   16 May 08  Andy Frank  Creation
//

using System.Text;
using System.Collections;

namespace Fan.Sys
{
  /// <summary>
  /// MimeType represents the parsed value of a Content-Type
  /// header per RFC 2045 section 5.1.
  /// </summary>
  public sealed class MimeType : FanObj
  {

  //////////////////////////////////////////////////////////////////////////
  // Construction
  //////////////////////////////////////////////////////////////////////////

    public static MimeType fromStr(string s) { return fromStr(s, true); }
    public static MimeType fromStr(string s, bool check)
    {
      try
      {
        // common interned mime types
        switch (s[0])
        {
          case 'i':
            if (s == "image/png")  return m_imagePng;
            if (s == "image/jpeg") return m_imageJpeg;
            if (s == "image/gif")  return m_imageGif;
            break;
          case 't':
            if (s == "text/plain") return m_textPlain;
            if (s == "text/html")  return m_textHtml;
            if (s == "text/xml")   return m_textXml;
            if (s == "text/plain; charset=utf-8") return m_textPlainUtf8;
            if (s == "text/html; charset=utf-8")  return m_textHtmlUtf8;
            if (s == "text/xml; charset=utf-8")   return m_textXmlUtf8;
            break;
          case 'x':
            if (s == "x-directory/normal") return m_dir;
            break;
        }

        return parse(s);
      }
      catch (ParseErr.Val e)
      {
        if (!check) return null;
        throw e;
      }
      catch (System.Exception)
      {
        if (!check) return null;
        throw ParseErr.make("MimeType",  s).val;
      }
    }

    private static MimeType parse(string s)
    {
      int slash = s.IndexOf('/');
      string media = s.Substring(0, slash);
      string sub = s.Substring(slash+1);
      Map pars = emptyParams();

      int semi = sub.IndexOf(';');
      if (semi > 0)
      {
        pars = doParseParams(sub, semi+1);
        sub = sub.Substring(0, semi).Trim();
      }

      MimeType r    = new MimeType();
      r.m_str       = s;
      r.m_mediaType = FanStr.lower(media);
      r.m_subType   = FanStr.lower(sub);
      r.m_params    = pars.ro();
      return r;
    }

    public static Map parseParams(string s) { return parseParams(s, true); }
    public static Map parseParams(string s, bool check)
    {
      try
      {
        return doParseParams(s, 0);
      }
      catch (ParseErr.Val e)
      {
        if (!check) return null;
        throw e;
      }
      catch (System.Exception)
      {
        if (!check) return null;
        throw ParseErr.make("MimeType params",  s).val;
      }
    }

    private static Map doParseParams(string s, int offset)
    {
      Map pars = new Map(Sys.StrType, Sys.StrType);
      pars.caseInsensitive(true);
      bool inQuotes = false;
      int keyStart = offset;
      int valStart = -1;
      int valEnd   = -1;
      int eq       = -1;
      bool hasEsc  = false;
      for (int i = keyStart; i<s.Length; ++i)
      {
        int c = s[i];

        if (c == '(' && !inQuotes)
          throw ParseErr.make("MimeType", s, "comments not supported").val;

        if (c == '=' && !inQuotes)
        {
          eq = i++;
          while (FanInt.isSpace(s[i])) ++i;
          if (s[i] == '"') { inQuotes = true; ++i; }
          else inQuotes = false;
          valStart = i;
        }

        if (eq < 0) continue;

        if (c == '\\' && inQuotes)
        {
          ++i;
          hasEsc = true;
          continue;
        }

        if (c == '"' && inQuotes)
        {
          valEnd = i-1;
          inQuotes = false;
        }

        if (c == ';' && !inQuotes)
        {
          if (valEnd < 0) valEnd = i-1;
          string key = s.Substring(keyStart, eq-keyStart).Trim();
          string val = s.Substring(valStart, valEnd+1-valStart).Trim();
          if (hasEsc) val = unescape(val);
          pars.set(key, val);
          keyStart = i+1;
          eq = valStart = valEnd = -1;
          hasEsc = false;
        }
      }

      if (keyStart < s.Length)
      {
        if (valEnd < 0) valEnd = s.Length-1;
        string key = s.Substring(keyStart, eq-keyStart).Trim();
        string val = s.Substring(valStart, valEnd+1-valStart).Trim();
        if (hasEsc) val = unescape(val);
        pars.set(key, val);
      }

      return pars;
    }

    private static string unescape(string s)
    {
      StringBuilder buf = new StringBuilder(s.Length);
      for (int i=0; i<s.Length; ++i)
      {
        int c = s[i];
        if (c != '\\') buf.Append((char)c);
        else if (s[i+1] == '\\') { buf.Append('\\'); i++; }
      }
      return buf.ToString();
    }

  //////////////////////////////////////////////////////////////////////////
  // Extension
  //////////////////////////////////////////////////////////////////////////

    public static MimeType forExt(string s)
    {
      if (s == null) return null;
      try
      {
        string val = (string)Sys.m_sysPod.props(m_etcUri, Duration.m_oneMin).get(FanStr.lower(s));
        if (val == null) return null;
        return MimeType.fromStr(val);
      }
      catch (System.Exception e)
      {
        System.Console.WriteLine("MimeType.forExt: " + s);
        Err.dumpStack(e);
        return null;
      }
    }

    static readonly Uri m_etcUri = Uri.fromStr("ext2mime.props");

  //////////////////////////////////////////////////////////////////////////
  // Identity
  //////////////////////////////////////////////////////////////////////////

    public override bool Equals(object obj)
    {
      if (!(obj is MimeType)) return false;
      MimeType x = (MimeType)obj;
      return
        m_mediaType == x.m_mediaType &&
        m_subType == x.m_subType &&
        m_params.Equals(x.m_params);
    }

    public override int GetHashCode()
    {
      return m_mediaType.GetHashCode() ^
             m_subType.GetHashCode() ^
             m_params.GetHashCode();
    }

    public override long hash()
    {
      return GetHashCode();
    }

    public override string toStr()
    {
      return m_str;
    }

    public override Type @typeof()
    {
      return Sys.MimeTypeType;
    }

  //////////////////////////////////////////////////////////////////////////
  // Methods
  //////////////////////////////////////////////////////////////////////////

    public string mediaType()
    {
      return m_mediaType;
    }

    public string subType()
    {
      return m_subType;
    }

    public Map @params()
    {
      return m_params;
    }

    public Charset charset()
    {
      string s = (string)m_params.get("charset");
      if (s == null) return Charset.utf8();
      return Charset.fromStr(s);
    }

  //////////////////////////////////////////////////////////////////////////
  // Lazy Load
  //////////////////////////////////////////////////////////////////////////

    internal static Map emptyParams()
    {
      Map q = m_emptyQuery;
      if (q == null)
      {
        q = new Map(Sys.StrType, Sys.StrType);
        q.caseInsensitive(true);
        q = (Map)q.toImmutable();
        m_emptyQuery = q;
      }
      return q;
    }
    static Map m_emptyQuery;

  //////////////////////////////////////////////////////////////////////////
  // Fields
  //////////////////////////////////////////////////////////////////////////

    internal static MimeType m_imagePng   = parse("image/png");
    internal static MimeType m_imageGif   = parse("image/gif");
    internal static MimeType m_imageJpeg  = parse("image/jpeg");
    internal static MimeType m_textPlain  = parse("text/plain");
    internal static MimeType m_textHtml   = parse("text/html");
    internal static MimeType m_textXml    = parse("text/xml");
    internal static MimeType m_dir        = parse("x-directory/normal");
    internal static MimeType m_textPlainUtf8 = parse("text/plain; charset=utf-8");
    internal static MimeType m_textHtmlUtf8  = parse("text/html; charset=utf-8");
    internal static MimeType m_textXmlUtf8   = parse("text/xml; charset=utf-8");

    private string m_mediaType;
    private string m_subType;
    private Map m_params;
    private string m_str;

  }
}